rfcs/20210827-benchdnn-model-support/README.md

```diff
@@ -0,0 +1,69 @@
+This allows to know a problem performance in "vacuum" environment - unrealistic
+conditions for end users. It still shows how efficient the kernel is, this is
+still an important part, but final performance values are not limited to kernel
+efficiency in "vacuum" environment only. Each next generation of hardware brings
+more compute elements and ususally do not improve a memory component which means
+that more and more problems become memory-bound rather than compute bound.
+Final performance users may see depends on many different factors.
+
+The problem of only existing approach that the result should be somehow scaled
+down to be closer to the "reality" but it is hard to say how much and how to
+estimate this ratio. Also certain performance problems cannot be found with such
+methodology. I.e., cores trash-trafficking effect due to different blocking
+strategies in adjacent primitives, or data eviction from caches due to different
+sets of weights, while a single problem fits well. It may also happen that HW
+features may change behavior in time. For example, compute frequency changes
+triggered by using some compute instructions (Intel AVX-512, Intel AMX, or
+DPAS).
+
+All these or other problems are hard or impossible to observe with existing
+benchdnn performance measuring toolset. That is why, developers or users who
+report issues have to use framework integration of oneDNN and submit trackers
+using this heavy-weighted instrument. There are several down sides of this
+approach:
+1) It requires frameworks infrastructure to be enabled on developer side.
+   Running different scripts is not that bad as trying to modify codes or other
+   environment around. It usually requires consultation from other teams or
+   reporter.
+2) To validate something new, especially API, it requires integration to be
+   implemented on framework side. It takes some time to make it done, which
+   postpones validation and may result in milestones misses when it comes to
+   new hardware enabling, delivering certain level of performance, etc.
+3) When it comes to performance measurements and analysis, it may be difficult
+   to distinguish a problem on framework side (integration code may have issues)
+   from the problem on the library side. In addition, there are certain
+   overheads that should be accounted when doing analysis.
+
+All these issues create a need to extend benchdnn performance measurements
+capabilities.
+
+## Proposal
+There are two general approaches that could be taken in order to improve a
+toolset:
+* "cold cache" mechanism. When executing a hot loop it would evict data from
+  cache(s) after each execution making new iteration to move data from RAM or
+  higher level cache(s).
+
+  Open questions:
+  - Not clear which memories to flush if there are two or more input tensors.
+  - Not clear what should be a level of eviction.
+  - What UI to provide to handle cons with enough level of flexibility?
```

Comment:
Regarding the "cold cache" my two cents based on debugging related performance issues.
In the past, I simulated it using multiple copies of memory, so that each iteration would work on memory different from previous, forcing to stream data from higher level memory hierarchy.

Some mix of pseudocode and code from my past hack
```
// *_scaling are inputs from user with new knobs.
// something like --<knob>=[float_multiplier][{B,L1,L2,L3}]
size_t src_mem_scaling, wei_mem_scaling, dst_mem_scaling;

// create copies of src_mem
const size src_size = src_md.size();
vector<mem> src_mems;
for (size_t cur_src_size = 0; cur_src_size < max(src_size, src_mem_scaling); cur_src_size += src_size) {
    src_mems.push_back(mem(src_md));
}
;;; ditto for wei, dst

// populate args
vector<args_t> args;
int i_s=0, i_w=0; i_d=0;
while (i_s != src.size() && i_w != wei.size() && i_d != dst.size()) {
    args_t tmp_args;
    i_s = i_s % src.size();
    tmp_args[SRC]=src_mems[i_s++];
    ;;; ditto wei, dst
    args.push_back(tmp);
}

// some plumbing and then finally
inline int measure_perf_individual(benchdnn_timer_t &t, dnnl_stream_t stream,
        perf_function_t &perf_func,
        std::vector<std::vector<dnnl_exec_arg_t>> &dnnl_args) {
    int i = 0;
    int n = dnnl_args.size();
    t.reset();
    while (true) {
        DNN_SAFE(perf_func(stream, dnnl_args[i]), WARN);
        t.stamp();
        if (should_stop(t)) break;
        if (++i >= n) i = 0;
    }
    return OK;
}
```
It is not perfect and granularity is limited by corresponding memory descriptors. But I think it can be used as a tool to debug.